driver/src/main/com/mongodb/util/JSONCallback.java

```diff
@@ -125,7 +125,7 @@ public Object objectDone() {
         } else if (b.containsField("$uuid")) {
             o = UUID.fromString((String) b.get("$uuid"));
         } else if (b.containsField("$binary")) {
-            int type = (Integer) b.get("$type");
+            int type = (b.get("$type") instanceof String) ? Integer.decode((String) b.get("$type")) : (Integer) b.get("$type");
```

Comment:
This is incorrect because when `$type` is a String it represents a hex string and not an int string.

It needs to be updated to: `Integer.valueOf((String) b.get("$type"), 16)`